### PR TITLE
Add the ability to auto-determine the presenter

### PR DESCRIPTION
**Pull request by drjermy**
I wonder whether you'd be interested in considering this pull-request. 

The idea here is to automatically use a presenter based on naming convention. So, in the User Model, the UserPresenter is loaded without needing to specify it in the $presenters array.

It also guesses the name for a named presenter, e.g. $users->present('certificate')->title would load the UserCertificatePresenter.

### CODE REVIEW

**Comment by ousid** [comment]
Hi, @drjermy 

Sorry for the late response/review. I was quite busy.

This is an interesting PR to merge, could you please elaborate more about the idea with examples?

Thanks!

**Comment by drjermy** [comment]
Currently when using Presenters, we setup by:
- implement CanPresent
- use UsesPresenters

And then, have to add a $presenters array that defines the Presenters. In most cases, we define a single default presenter. 

So, rather than doing this, why not automatically look for a Presenter in a standard location (Where php artisan make:presenter puts a presenter) with the name "ModelPresenter" where Model is the model name.

If $presenters is set, then it overwrites the defaults.

So, if I had a model called User and implemented CanPresent and added UsesPresenters, then the default presenter would autoload UserPresenter.php which I'd created using php artisan make:presenter

**Comment by drjermy** [comment]
@ousid any update?

**Comment by ousid** [comment]
> @ousid any update?


Hey,  I'm seeing failed tests, can you please take a look and see what's wrong exactly?

Thanks!

File: src/Concerns/UsesPresenters.php [hunk 1]
```diff
@@ -9,18 +9,62 @@
...
-     * @return mixed
+     * @throws PresenterException
      */
-    public function present(string $type = 'default')
+    public function present(string $type = 'default'): mixed
     {
-        if (array_key_exists($type, $this->presenters)) {
-            return new $this->presenters[$type]($this);
+        $presenter = $this->getDefaultPresenterName($type);
+
```
Comment (ousid) [inline src/Concerns/UsesPresenters.php:29]: I see a __Code smell__ in here, consider using aggressive condition statement, to clean the code, and understandable.

```suggestion
if (
    is_array($this->presenters) &&
    array_key_exists($type, $this->presenters ?? [])
) {
  $presenter =$this->presetners[$type];
}

if (is_string($this->presenter ?? null)) {
  $presenter = $this->presenter;
}

if(class_exists($presenter)) {
   return new $presenter($this);
}

throw new PresenterException();
```

File: tests/Presenters/ItemPresenter.php [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+;
+
+use Coderflex\LaravelPresenter\Presenter;
+
+class ItemPresenter extends Presenter
+{
+    public function slug()
+    {
+        return str($this->model->title)->slug()->toString();
```
Comment (ousid) [inline tests/Presenters/ItemPresenter.php:13]: Consider using `Str` facade.

File: tests/Presenters/ItemSubdomainPresenter.php [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+;
+
+use Coderflex\LaravelPresenter\Presenter;
+
+class ItemSubdomainPresenter extends Presenter
+{
+    public function caps()
+    {
+        return str($this->model->title)->upper()->toString();
```
Comment (ousid) [inline tests/Presenters/ItemSubdomainPresenter.php:13]: Same here, use the `Str` facade, or invoke the `str` function on the namespaces.

File: README.md [hunk 4]
```diff
@@ -111,7 +111,7 @@ If you want to change the directory, you have two options.
 First options is to set the full namespace while you're creating the presenter class
 
 ```bash
-php artisna presneter:make App\Models\Presenter\UserPresenter
+php artisan presneter:make App\Models\Presenter\UserPresenter
```
Comment (ousid) [inline README.md:114]: Maybe you want to create an alias for the command? Something like `php artisan make:presenter`, without removing the current?

**Review by ousid — CHANGES_REQUESTED** [review]
@drjermy 
Please take a look at the reviews, and consider fix the failing tests (most of them relate to the `str` method)

File: README.md (hunk 4 above)
Comment (drjermy) [reply README.md:114]: This was just a spelling error change.

**Comment by ousid** [comment]
Tests still failing, could you please run them locally, and once all passes, tag me for a review.

Thanks!

**Comment by ousid** [comment]
Closing this PR, because of the failing tests.

**Pull request closed without merging**